Source/Core/AsyncQueue.cs

```diff
@@ -0,0 +1,70 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Boogie;
+
+/// <summary>
+/// A queue from which items can be dequeued even if they have not been enqueued yet.
+/// All methods are thread-safe.
+/// Also supports pushing elements to the front of the queue.
+/// </summary>
+public class AsyncQueue<T>
+{
+  private readonly LinkedList<T> items = new();
+  private readonly Queue<TaskCompletionSource<T>> customers = new();
+
+  public void Enqueue(T value)
+  {
+    lock (this) {
+      while (customers.TryDequeue(out var customer)) {
+        if (customer.TrySetResult(value)) {
```

Comment:
When can this be false?

Reply:
When the customer Task has been cancelled

Reply:
Gosh, good catch.